src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt

```diff
@@ -671,20 +675,170 @@ data class ImportContext private constructor(
+    }
+
+    /**
+     * Traverse types in `elemTy` and collects all items that unresolved in current context.
+     */
+    private fun getTypeReferencesInfoFromTy(
+        context: RsElement,
+        elemTy: Ty,
+        useAliases: Boolean
+    ): TypeReferencesInfo = getTypeReferencesInfo(context, listOf(elemTy)) { ty, result ->
+        collectImportSubjectsFromTy(ty, emptySubstitution, result, useAliases)
+    }
+
+    private fun <T> getTypeReferencesInfo(
+        context: RsElement,
+        elements: Collection<T>,
+        collector: (T, MutableSet<RsQualifiedNamedElement>) -> Unit
+    ): TypeReferencesInfo {
+        val result = hashSetOf<RsQualifiedNamedElement>()
+        elements.forEach { collector(it, result) }
+        return processRawImportSubjects(context, result)
+    }
+
+    private fun collectImportSubjectsFromTypeReferences(
+        context: RsElement,
+        subst: Substitution,
+        result: MutableSet<RsQualifiedNamedElement>,
+        useAliases: Boolean
+    ) {
+        context.accept(object : RsVisitor() {
+            override fun visitTypeReference(reference: RsTypeReference) =
+                collectImportSubjectsFromTy(reference.type, subst, result, useAliases)
+
+            override fun visitElement(element: RsElement) =
+                element.acceptChildren(this)
+        })
+    }
+
+    private fun collectImportSubjectsFromTy(
+        ty: Ty,
+        subst: Substitution,
+        result: MutableSet<RsQualifiedNamedElement>,
+        useAliases: Boolean
+    ) {
+        ty.substitute(subst).visitWith(object : TypeVisitor {
+            override fun visitTy(ty: Ty): Boolean {
+                when (ty) {
+                    is TyAdt -> {
+                        val alias = ty.aliasedBy?.element.takeIf { useAliases } as? RsQualifiedNamedElement
+                        result += alias ?: ty.item
+                    }
+                    is TyAnon -> result += ty.traits.map { it.element }
+                    is TyTraitObject -> result += ty.trait.element
+                    is TyProjection -> {
+                        result += ty.trait.element
+                        result += ty.target
+                    }
+                }
+                return ty.superVisitWith(this)
+            }
+        })
+    }
+
+    /**
+     * Takes `rawImportSubjects` and filters items that are unresolved in the current `context`.
+     * Then splits the items into two sets:
+     * - Items that should be imported
+     * - Items that can't be imported
+     */
+    private fun processRawImportSubjects(
+        context: RsElement,
+        rawImportSubjects: Set<RsQualifiedNamedElement>
+    ): TypeReferencesInfo {
+        val importSubjects = hashMapOf<String, MutableSet<RsQualifiedNamedElement>>()
+        for (element in rawImportSubjects) {
+            val name = element.name ?: continue
+            importSubjects.getOrPut(name, ::hashSetOf) += element
+        }
```

Comment:
Could you come up with case when value set has size > 1?

Reply:
```
use crate::a::foo;

mod a {
    mod b { pub struct S; }
    mod c { pub struct S; }
    pub fn foo() -> (b::S, c::S) { unimplemented!() }
}

fn main() {
    let x/*caret*/ = foo();
}
```